File: python/CAPSTONE_2/pages/cart_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support import expected_conditions as EC
from pages.base_page import BasePage
import time

class CartPage(BasePage):
    def __init__(self, driver):
        super().__init__(driver)
        self.product_items = (By.CSS_SELECTOR, "div.shelf-item")
        self.add_to_cart_buttons = (By.CSS_SELECTOR, "div.shelf-item__buy-btn")
        self.cart_icon = (By.CSS_SELECTOR, "div.float-cart__toggle")
        self.cart_items = (By.CSS_SELECTOR, "div.float-cart__shelf-container div.shelf-item")

    def add_multiple_items(self, count=3):
        """Add multiple visible items to the cart."""
        print("⏳ Waiting for product grid before adding to cart...")
        self.wait.until(EC.presence_of_all_elements_located(self.product_items))
        buttons = self.wait.until(EC.presence_of_all_elements_located(self.add_to_cart_buttons))
        print(f"✅ Found {len(buttons)} 'Add to cart' buttons.")

        for i, button in enumerate(buttons[:count]):
            self.driver.execute_script("arguments[0].scrollIntoView(true);", button)
            self.driver.execute_script("window.scrollBy(0, 100);")
            self.driver.execute_script("arguments[0].click();", button)
            print(f"🛒 Added item {i+1} to cart.")
            time.sleep(1)
        print(f"✅ Added {count} products to cart successfully.")

    def verify_cart_items(self):
        """Open cart and count number of items."""
        print("🧾 Opening cart to verify items...")

        try:
            possible_cart_locators = [
                (By.CSS_SELECTOR, "div.float-cart__toggle"),
                (By.CSS_SELECTOR, "span.bag--float-cart-closed"),
                (By.XPATH, "//div[contains(@class,'float-cart') and not(contains(@class,'open'))]")
            ]

            cart_icon = None
            for locator in possible_cart_locators:
                try:
                    cart_icon = self.wait.until(EC.presence_of_element_located(locator))
                    if cart_icon:
                        print(f"✅ Found cart icon using locator: {locator}")
                        break
                except Exception:
                    continue

            if not cart_icon:
                raise Exception("❌ Could not find cart icon even after checking multiple locators.")

            # Scroll to top (cart usually top-right)
            self.driver.execute_script("window.scrollTo(0, 0);")
            time.sleep(1)

            # Click via JS
            self.driver.execute_script("arguments[0].click();", cart_icon)
            print("🛍️ Clicked cart icon successfully.")
            time.sleep(2)

            # Wait for cart items to appear
            items = self.wait.until(EC.presence_of_all_elements_located(self.cart_items))
            print(f"🛒 Cart contains {len(items)} item(s).")
            return len(items)

        except Exception as e:
            print(f"❌ Could not open cart or verify items: {e}")
            self.driver.save_screenshot("cart_debug.png")
            raise

File: selenium with java/testcase3/CAPSTONE_2 (JAVA)/src/main/java/pages/CartPage.java
package pages;

import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.ExpectedConditions;
import java.util.List;

public class CartPage extends BasePage {

    private By productItems = By.cssSelector("div.shelf-item");
    private By addToCartButtons = By.cssSelector("div.shelf-item__buy-btn");
    private By cartToggle = By.xpath("//div[contains(@class,'float-cart__toggle')]");
    private By cartContainer = By.xpath("//div[contains(@class,'float-cart') and contains(@class,'open')]");
    private By cartCloseButton = By.cssSelector("div.float-cart__close-btn");
    private By cartItems = By.cssSelector("div.float-cart__shelf-container div.shelf-item");

    public CartPage(WebDriver driver) {
        super(driver);
    }

    // 🔹 Add multiple items to cart
    public void addMultipleItems(int count) {
        try {
            wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(productItems));
            List<WebElement> buttons = driver.findElements(addToCartButtons);
            System.out.println("✅ Found " + buttons.size() + " 'Add to cart' buttons.");

            for (int i = 0; i < count && i < buttons.size(); i++) {
                WebElement btn = buttons.get(i);
                wait.until(ExpectedConditions.elementToBeClickable(btn)).click();
                System.out.println("🛒 Added item " + (i + 1) + " to cart.");
                Thread.sleep(1000);
            }
            System.out.println("✅ Added " + count + " products to cart successfully.");

        } catch (Exception e) {
            System.err.println("⚠️ Error adding products: " + e.getMessage());
        }
    }

    // 🔹 Verify cart contents
    public int verifyCartItems() {
        try {
            System.out.println("🧾 Checking cart state...");

            // Scroll up to make sure floating cart is visible
            ((JavascriptExecutor) driver).executeScript("window.scrollTo(0, 0);");
            Thread.sleep(1000);

            // If cart is already open, no need to click toggle
            boolean cartAlreadyOpen = driver.findElements(cartContainer).size() > 0;
            if (!cartAlreadyOpen) {
                WebElement toggle = wait.until(ExpectedConditions.elementToBeClickable(cartToggle));
                toggle.click();
                System.out.println("🛍️ Cart opened successfully.");
            } else {
                System.out.println("🛍️ Cart is already open — skipping toggle click.");
            }

            // Wait for cart items to load
            List<WebElement> items = wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(cartItems));
            System.out.println("🛒 Cart contains " + items.size() + " item(s).");

            // Optional: close the cart for next test
            if (driver.findElements(cartCloseButton).size() > 0) {
                driver.findElement(cartCloseButton).click();
                System.out.println("❎ Closed cart after verification.");
            }

            return items.size();

        } catch (Exception e) {
            System.err.println("❌ Failed to verify cart: " + e.getMessage());
            return 0;
        }
    }
}
